### PR TITLE
tests: parameterize remote store

### PR DESCRIPTION
**Pull request by fgimenez**
The remote store to be used (currently production or staging) is defined by an external environment variable `SPREAD_REMOTE_STORE` (defaults to production). Also, the store variants for production have been renamed to remote.

### CODE REVIEW

**Comment by mvo5** [comment]
The infamous `error: cannot refresh "test-snapd-tools": cannot get refresh information for snap "test-snapd-tools": Post https://search.apps.ubuntu.com/api/v1/snaps/metadata: EOF` again in the tests.

File: tests/lib/store.sh [hunk 1]
```diff
@@ -15,57 +19,45 @@ EOF
...
+
     mkdir -p $top_dir/asserts
     # debugging
     systemctl status fakestore || true
     echo "Given a controlled store service is up"
 
     https_proxy=${https_proxy:-}
     http_proxy=${http_proxy:-}
-    systemd_create_and_start_unit fakestore "$(which fakestore) -start -dir $top_dir -addr localhost:11028 -https-proxy=${https_proxy} -http-proxy=${http_proxy} $@" "SNAPD_DEBUG=1 SNAPD_DEBUG_HTTP=7"
+    systemd_create_and_start_unit fakestore "$(which fakestore) -start -dir $top_dir -addr localhost:11028 -https-proxy=${https_proxy} -http-proxy=${http_proxy} -assert-fallback" "SNAPD_DEBUG=1 SNAPD_DEBUG_HTTP=7"
```
Comment (pedronis) [inline tests/lib/store.sh]: here when we run fakestore we also need to pass to it SNAPPY_USE_STAGING_STORE=1 if we are using staging

File: tests/main/refresh-devmode/task.yaml [hunk 2]
```diff
@@ -13,35 +13,33 @@ environment:
...
-    setup_store $STORE_TYPE $BLOB_DIR
+    echo "Given a snap is installed"
+    snap install --devmode test-snapd-tools
 
     if [ "$STORE_TYPE" = "fake" ]; then
+        . $TESTSLIB/store.sh
+        setup_fake_store $BLOB_DIR
+
         echo "And a new version of that snap put in the controlled store"
         fakestore -dir $BLOB_DIR -make-refreshable test-snapd-tools
```
Comment (pedronis) [inline tests/main/refresh-devmode/task.yaml]: also these usages of fakestore need SNAPPY_USE_STAGING_STORE=1

**Comment by pedronis** [comment]
not 100% sure those changes will fix things but should go in the right direction

**Comment by fgimenez** [comment]
hey @pedronis thanks a lot for your review! I've just pushed changes to add the `-tags withstagingkeys` option when building fakestore for staging and also adding `SNAPPY_USE_STAGING_STORE=1` when setting it up. 

With that in place I was getting this kind of errors in the `ubuntu-16.04-64` (classic) system:
```
2016/12/14 10:27:50 Error executing linode:ubuntu-16.04-64:tests/main/revert-devmode:fake : 
-----
+ [[ fake = \f\a\k\e ]]
+ [[ ubuntu-16.04-64 =~ ubuntu-core-16-* ]]
+ echo 'When a refresh is made'
When a refresh is made
+ snap refresh --devmode --edge test-snapd-tools
error: cannot refresh "test-snapd-tools": cannot refresh snap-declaration for "core": snap-declaration (99T7MUlRhtI3U0QFgl5mXXESAiSwt776; series:16) not found
```

That core snap id seems to be coming yet from production. Not sure if it is related to these errors but I realized that the core snap was being installed before the switch to staging (so it was installed from production), and I've added also changes for switching to staging before any snap is installed. However now I'm getting this error on the first snap install command:
```
+ snap install test-snapd-tools
error: cannot perform the following tasks:
- Fetch and check assertions for snap "core" (6) (internal error: circular assertions are not expected: account (canonical))
```

The current changeset produce this ^ error. Please have a look when you have the time.

Thanks! :)

File: spread.yaml [hunk 3]
```diff
@@ -230,7 +231,10 @@ prepare: |
     go get ./tests/lib/snapbuild
 
     # Build fakestore.
-    go get ./tests/lib/fakestore/cmd/fakestore
+    if [ "$REMOTE_STORE" = staging ]; then
+        fakestore_tags="-tags withstagingkeys"
+    fi
+    go get $fakestore_tags ./tests/lib/fakestore/cmd/fakestore
```
Comment (pedronis) [inline spread.yaml]: this gives  /bin/bash: line 150: fakestore_tags: unbound variable if REMOTE_STORE is not staging

**Comment by pedronis** [comment]
to progress further we need #2505 

to fix the issue that SNAPPY_USE_STAGING_STORE=1 must be set also when we call fakestore -make-refreshable

and we need this further tweak:

   --- a/tests/lib/store.sh
   +++ b/tests/lib/store.sh
   @@ -50,7 +50,7 @@ setup_fake_store(){
        systemd_create_and_start_unit fakestore "$(which fakestore) -start -dir $top_dir -addr localhost:11028 -https-proxy=${https_proxy} -http-proxy=${http_proxy} -assert-fallback" "SNAPD_DEBUG=1 SNAPD_DEBUG_HTTP=7 $fakestore_mark"
 
        echo "And snapd is configured to use the controlled store"
    -    _configure_store_backends "SNAPPY_FORCE_CPI_URL=http://localhost:11028"   "SNAPPY_FORCE_SAS_URL=http://localhost:11028"
    +    _configure_store_backends "SNAPPY_FORCE_CPI_URL=http://localhost:11028" "SNAPPY_FORCE_SAS_URL=http://localhost:11028 $fakestore_mark"
    }
 
    teardown_fake_store(){

**Comment by fgimenez** [comment]
@pedronis thanks a lot for your review, comments addressed. Please take look, especially the `init_fakestore` helper in `tests/lib/store.sh`.

File: tests/lib/store.sh [hunk 4]
```diff
@@ -15,57 +19,63 @@ EOF
...
+teardown_staging_store(){
+    if systemctl status snapd.socket; then
+        systemctl stop snapd.socket
+    fi
+    rm -rf $STORE_CONFIG
+    systemctl daemon-reload
+    systemctl start snapd.socket
+}
+
+init_fakestore(){
```
Comment (pedronis) [inline tests/lib/store.sh]: not sure it's clear the difference between setup_fake_store and init_fakestore, can we call this one 

init_fake_refreshes ?

File: spread.yaml [hunk 5]
```diff
@@ -263,6 +268,11 @@ suites:
                 prepare_classic
             fi
         restore: |
+            if [ "$REMOTE_STORE" = staging ]; then
+                . $TESTSLIB/store.sh
+                teardown_staging_store
+            fi
+
             $TESTSLIB/reset.sh
```
Comment (pedronis) [inline spread.yaml]: can't we change this somehow to  "reset.sh --store"  and move the lines above inside reset.sh ? the asymmetry with prepare (which does the setup_staging_store) is a bit strange to me

File: tests/lib/store.sh (hunk 4 above)
Comment (fgimenez) [inline tests/lib/store.sh]: sure, done

File: spread.yaml (hunk 5 above)
Comment (fgimenez) [inline spread.yaml]: Indeed, we handle a bit different prepare and restore (prepare sources and calls functions, restore just calls the executable reset.sh file and the functions are called from inside), I've just moved the store teardown to reset.sh, plz have a look.

**Comment by fgimenez** [comment]
Thanks @pedronis comments addressed

File: tests/lib/fakestore/store/store.go [hunk 6]
```diff
@@ -365,7 +374,11 @@ func (s *Store) bulkEndpoint(w http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	snapIDtoName, err := addSnapIDs(bs, someSnapIDtoName)
+	remoteStore := os.Getenv("REMOTE_STORE")
```
Comment (pedronis) [inline tests/lib/fakestore/store/store.go]: as we discussed we can/should use SNAPPY_USE_STAGING_STORE == "1" here instead

Comment (chipaca) [inline tests/lib/fakestore/store/store.go]: `osutil.GetenvBool`

Comment (fgimenez) [inline tests/lib/fakestore/store/store.go]: @chipaca supernice thanks!

File: tests/main/local-install-w-metadata/task.yaml [hunk 7]
```diff
@@ -11,11 +11,20 @@ execute: |
...
     # XXX snap download should do this as well
-    curl -H "Accept: application/x.ubuntu.assertion" -o snap-decl.assertion https://assertions.ubuntu.com/v1/assertions/snap-declaration/16/eFe8BTR5L5V9F7yHeMAPxkEr2NdUXMtw
+    curl -H "Accept: application/x.ubuntu.assertion" -o snap-decl.assertion https://${assertions_server}/v1/assertions/snap-declaration/16/${snap_id}
     # XXX have a 'snap snap-digest' command?
     digest=$(go run digest.go test-snapd-tools_*.snap)
-    curl -H "Accept: application/x.ubuntu.assertion" -o snap-rev.assertion https://assertions.ubuntu.com/v1/assertions/snap-revision/${digest}
+    curl -H "Accept: application/x.ubuntu.assertion" -o snap-rev.assertion https://${assertions_server}/v1/assertions/snap-revision/${digest}
     # add them
     snap ack snap-decl.assertion
     snap ack snap-rev.assertion
```
Comment (pedronis) [inline tests/main/local-install-w-metadata/task.yaml]: we don't need all of this anymore nowadays:  snap download creates a .assert file

so we just need:   snap ack test-snapd-tool_*.assert

Comment (pedronis) [inline tests/main/local-install-w-metadata/task.yaml]: otoh the "snap download" needs SNAPPY_USE_STAGING=1 I think

**Comment by fgimenez** [comment]
thanks a lot @pedronis, done. I've taken the opportunity to define a toplevel `SNAPPY_USE_STAGING_STORE`, which uncovered more missing snaps in staging (pi2 and classic), and a new error:
```
2016/12/21 10:47:23 Error executing linode:ubuntu-16.04-64:tests/main/known-remote : 
-----
+ echo 'Check getting assertion from the store'
Check getting assertion from the store
++ snap known --remote model series=16 brand-id=canonical model=pi2
error: model (pi2; series:16 brand-id:canonical) not found
```
Only this one and snap-info failing on `ubuntu-16.04-64`

**Comment by fgimenez** [comment]
With all the required snaps in place and the latest fixes for prepare-image-uboot the only test failing on `ubuntu-16.04-64` is `tests/main/known-remote`.

**Comment by fgimenez** [comment]
With the latest commits and the changes on the store side the suite passes on both classic systems and only `tests/main/ubuntu-core-device-reg` fails on the ubuntu-core system because of errors during the device serial request process:

```
root@localhost:~# snap changes
ID   Status  Spawn                 Ready                 Summary
1    Done    2017-01-10T15:53:47Z  2017-01-10T15:53:49Z  Initialize system state
2    Doing   2017-01-10T15:53:49Z  -                     Initialize device

root@localhost:~# snap change 2
Status  Spawn                 Ready                 Summary
Done    2017-01-10T15:53:49Z  2017-01-10T15:53:50Z  Run prepare-device hook
Done    2017-01-10T15:53:49Z  2017-01-10T15:53:53Z  Generate device key
Doing   2017-01-10T15:53:49Z  -                     Request device serial

......................................................................
Request device serial

2017-01-10T15:53:54Z ERROR cannot deliver device serial request: unexpected status 400
2017-01-10T16:05:21Z ERROR cannot deliver device serial request: unexpected status 400
2017-01-10T16:06:22Z ERROR cannot deliver device serial request: unexpected status 400
2017-01-10T16:07:23Z ERROR cannot deliver device serial request: unexpected status 400
2017-01-10T16:08:23Z ERROR cannot deliver device serial request: unexpected status 400
2017-01-10T16:09:24Z ERROR cannot deliver device serial request: unexpected status 400
2017-01-10T16:10:25Z ERROR cannot deliver device serial request: unexpected status 400
2017-01-10T16:11:25Z ERROR cannot deliver device serial request: unexpected status 400
2017-01-10T16:12:26Z ERROR cannot deliver device serial request: unexpected status 400
2017-01-10T16:13:27Z ERROR cannot deliver device serial request: unexpected status 400
```

This will be handled on a different branch.

**Comment by fgimenez** [comment]
All green after uploading the pc gadget snap from production.

File: tests/lib/store.sh [hunk 8]
```diff
@@ -4,8 +4,12 @@ STORE_CONFIG=/etc/systemd/system/snapd.service.d/store.conf
 . $TESTSLIB/systemd.sh
 
 _configure_store_backends(){
-    systemctl stop snapd.service snapd.socket
-    mkdir -p $(dirname $STORE_CONFIG)
+    for snap in snapd.service snapd.socket; do
+        if systemctl status "$snap"; then
```
Comment (chipaca) [inline tests/lib/store.sh]: Just call `systemctl stop snapd.service snapd.socket snapd.refresh.timer`. If any of them are running it'll stop them.

File: tests/lib/store.sh [hunk 9]
```diff
@@ -15,57 +19,53 @@ EOF
     systemctl start snapd.socket
 }
 
+setup_staging_store(){
+    _configure_store_backends "SNAPPY_USE_STAGING_STORE=1"
+}
+
+teardown_staging_store(){
+    if systemctl status snapd.socket; then
```
Comment (chipaca) [inline tests/lib/store.sh]: I think here again you just want to stop it.

File: tests/lib/store.sh (hunk 8 above)
Comment (fgimenez) [inline tests/lib/store.sh]: @chipaca nice, i thought it would gave error if already stopped, thanks!

**Pull request merged**